rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,309 @@
+tools updates](#self-managed-client-tools-updates), a series of warnings and
+errors with increasing urgency will be shown to the user.
+
+If the version of client tools is within the same major version as advertised
+by the cluster, a warning will be shown to urge the user to enroll in automatic
+updates. Warnings will not prevent the user from using client tools that are
+slightly out of date.
+
+```
+$ tsh login --proxy=proxy.example.com
+Warning: Client tools are out of date, update to vX.Y.Z.
+
+Update Teleport to vX.Y.Z from https://goteleport.com/download or your system
+package manager.
+
+Run "tsh autoupdate enroll" to enroll in automatic updates and keep client
+tools like tsh and tctl automatically updated.
+
+[...]
+```
+
+If the version of client tools is 1 major version below the version advertised
+by the cluster, a warning will be shown that indicates some functionality may
+not work.
+
+```
+$ tsh login --proxy=proxy.example.com
+WARNING: Client tools are 1 major version out of date, update to vX.Y.Z.
+
+Some functionality may not work. Update Teleport to vX.Y.Z from
+https://goteleport.com/download or your system package manager.
+
+Run "tsh autoupdate enroll" to enroll in automatic updates and keep client
+tools like tsh and tctl automatically updated.
+```
+
+If the version of client tools is 2 (or more) versions lower than the version
+advertised by the cluster or 1 (or more) version greater than the version
+advertised by the cluster, an error will be shown and will require the user to
+use the `--skip-version-check` flag.
+
+```
+$ tsh login --proxy=proxy.example.com
+ERROR: Client tools are N major versions out of date, update to vX.Y.Z.
+
+Some functionality will not work. Update Teleport to vX.Y.Z from
+https://goteleport.com/download or your system package manager.
+
+Run "tsh autoupdate enroll" to enroll in automatic updates and keep client
+tools like tsh and tctl automatically updated.
+
+Use the "--skip-version-check" flag to bypass this check and attempt to connect
+to this cluster.
+```
+
+#### Self-managed client tools updates
+
+Cluster administrators that want to self-manage client tools updates will be
+able to watch for changes to client tools versions which can then be used to
+trigger other integrations (using MDM software like Jamf) to update the
+installed version of client tools on endpoints.
+
+```
+$ tctl autoupdate watch
+{"tools_version": "1.0.0"}
+{"tools_version": "1.0.1"}
+{"tools_version": "2.0.0"}
+
+[...]
+```
+
+Cluster administrators who prefer to use the Teleport API can create a watcher
+on the `types.ClusterMaintenanceConfig` resource and trigger other integrations
+on changes to the `tools_version` field.
+
+```
+// Create a watcher to monitor for changes to client tools version.
+watch, err := p.TeleportClient.NewWatcher(ctx, types.Watch{
+    Kinds: []types.WatchKind{
+        types.WatchKind{Kind: types.ClusterMaintenanceConfig},
```

Comment:
Hmm, shouldn't this be watching for changes in actual version reported by ping endpoint, rather than for cluster maintenance config resource changes? Cause this will only "fire" when you update the resource, not when e.g. the cluster is updated, or am I missing something?

Reply:
Good point, I think I'll just remove this and leave `tctl autoupdate watch`.

Reply:
We don't have watchers for the AU endpoint, and it's answer might not be consistent across proxies currently.
The way the version server is currently working is via configuration at the proxy level. This means we have no cluster-wide version control. This was not intentional, this was a shortcut to implement the proxy version server faster.

I think we might not want to add yet another resource, but we could use the existing CMC to hold the version, maybe adding a `Status` field. 

```
kind: cluster_maintenance_config
spec:
  tools_version: auto | off | X.Y.Z
  [...]
status:
  default_agents_version: x.y.w
  default_tools_version: x.y.z
```

The status would be updated in the following cases:

- when you edit the spec (e.g. change from off to auto, or set a specific version)
- when a Teleport-Clould specific goroutine updates it (this way we can do pull-based if we want)
- when the auth server starts and `tools_version` is `auto` (can it check the lowest known proxy server version?)

This would allow all proxies to watch for updates and answer correctly.

Reply:
What about this? I'll do is add a helper function to our API that watches for changes to `types.ClusterMaintenanceConfig` and `types.Proxy` and sends the version if it changes over a channel. If `tools_version: X.Y.Z` this will just send changes from `types.ClusterMaintenanceConfig` and in `tools_version: auto` it will send changes when all proxies switch over to a updated version.

It can just keep track of state locally.